### PR TITLE
virtctl: Add option to wrap local scp client to scp command

### PR DESCRIPTION
**Pull request by 0xFelix**
**What this PR does / why we need it**:

This adds the option to use the local scp client instead of the native go
ssh client for scp operations. Like for the local ssh client virtctl is
used as ProxyCommand for scp.

**Release note**:

```release-note
Add option to wrap local scp client to scp command
```

### CODE REVIEW

File: pkg/virtctl/scp/native.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		return fmt.Errorf("local path %q does not exist, can't copy it", localPath)
+	}
+
+	if o.recursive {
+		if isFile {
+			return fmt.Errorf("local path %q is not a directory but '--recursive' was provided", localPath)
+		}
+
+		if err := client.CopyDirToRemote(localPath, remotePath, &scp.DirTransferOption{PreserveProp: o.preserve}); err != nil {
+			return err
```
Comment (alicefr) [inline pkg/virtctl/scp/native.go:51]: For simplify a bit the code you could return directly the result of the function `client.CopyDirToRemote`. If you return here you could also avoid the else and the last  ̇return nil ̇

File: pkg/virtctl/scp/native.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+		}
+	} else {
+		if isDir {
+			return fmt.Errorf("local path %q is a directory but '--recursive' was not provided", localPath)
+		}
+
+		if err := client.CopyFileToRemote(localPath, remotePath, &scp.FileTransferOption{PreserveProp: o.preserve}); err != nil {
+			return err
```
Comment (alicefr) [inline pkg/virtctl/scp/native.go:59]: Same here, simply return the function:
```golang 
 return client.CopyFileToRemote(localPath, remotePath, &scp.FileTransferOption{PreserveProp: o.preserve})
```

**Comment by alicefr** [comment]
@0xFelix do we have an option to pass additional options? I mean the option `-o ssh_option` in the scp command. Users might want to rely on the local scp command because they need to pass additional flags maybe not supported by the native version.

**Comment by 0xFelix** [comment]
@alicefr Yes, that is possible. The scp cmd takes the same flags as the ssh cmd. Additional options that should be passed to scp can be defined with `-t`.

The functests do that for example: https://github.com/kubevirt/kubevirt/blob/1d530ddb8f058406af15dbca7349c8346cb7d20e/tests/virtctl/scp.go#L50

File: pkg/virtctl/scp/native.go (hunk 1 above)
Comment (0xFelix) [reply pkg/virtctl/scp/native.go:51]: Obviously, I was sleeping here. :+1:

File: pkg/virtctl/scp/native.go (hunk 2 above)
Comment (0xFelix) [reply pkg/virtctl/scp/native.go:59]: Same :+1:

**Comment by alicefr** [comment]
Looks good
/lgtm

File: tests/virtctl/ssh.go [hunk 3]
```diff
@@ -74,12 +45,41 @@ var _ = Describe("[sig-compute][virtctl]SSH", func() {
 			"-t", "-o StrictHostKeyChecking=no",
 			"-t", "-o UserKnownHostsFile=/dev/null",
 			"--command", "true",
-			vmi.Name)
+			vmiName)
 		Expect(err).ToNot(HaveOccurred())
 		cmd.Env = append(cmd.Env,
 			"SSH_AUTH_SOCK=/dev/null")
 		out, err := cmd.CombinedOutput()
 		fmt.Println(string(out))
```
Comment (lyarwood) [inline tests/virtctl/ssh.go:53]: Can you assert that this isn't empty instead of printing it out?

File: tests/virtctl/ssh.go [hunk 4]
```diff
@@ -74,12 +45,41 @@ var _ = Describe("[sig-compute][virtctl]SSH", func() {
...
 		Expect(err).ToNot(HaveOccurred())
```
Comment (lyarwood) [inline tests/virtctl/ssh.go:54]: supernit - assert this first before attempting to capture any output?

File: tests/virtctl/scp.go [hunk 5]
```diff
@@ -23,6 +23,48 @@ var _ = Describe("[sig-compute][virtctl]SCP", func() {
...
+			args = append(args, "--recursive")
+		}
+		args = append(args, src, dst)
+
+		_, cmd, err := clientcmd.CreateCommandWithNS(util.NamespaceTestDefault, "virtctl", args...)
+		Expect(err).ToNot(HaveOccurred())
+		cmd.Env = append(cmd.Env,
+			"SSH_AUTH_SOCK=/dev/null")
+		out, err := cmd.CombinedOutput()
+		fmt.Println(string(out))
```
Comment (lyarwood) [inline tests/virtctl/scp.go:64]: As with the ssh tests, can you assert something about this and stop printing this out?

File: tests/virtctl/scp.go [hunk 6]
```diff
@@ -23,6 +23,48 @@ var _ = Describe("[sig-compute][virtctl]SCP", func() {
...
+			"-t", "-O",
+		}
+		if recursive {
+			args = append(args, "--recursive")
+		}
+		args = append(args, src, dst)
+
+		_, cmd, err := clientcmd.CreateCommandWithNS(util.NamespaceTestDefault, "virtctl", args...)
+		Expect(err).ToNot(HaveOccurred())
+		cmd.Env = append(cmd.Env,
```
Comment (lyarwood) [inline tests/virtctl/scp.go:61]: Is this required if we've already `Expect(os.Setenv("SSH_AUTH_SOCK", "/dev/null")).To(Succeed())` in `BeforeEach`?

File: tests/virtctl/ssh.go [hunk 7]
```diff
@@ -74,12 +45,41 @@ var _ = Describe("[sig-compute][virtctl]SSH", func() {
 			"-t", "-o StrictHostKeyChecking=no",
 			"-t", "-o UserKnownHostsFile=/dev/null",
 			"--command", "true",
-			vmi.Name)
+			vmiName)
 		Expect(err).ToNot(HaveOccurred())
 		cmd.Env = append(cmd.Env,
```
Comment (lyarwood) [inline tests/virtctl/ssh.go:50]: Is this required if we've already Expect(os.Setenv("SSH_AUTH_SOCK", "/dev/null")).To(Succeed()) in BeforeEach?

**Review by lyarwood — COMMENTED** [review]
This mostly LGTM but I've got a few questions and supernits in the functional tests if you can take a look.

File: tests/virtctl/ssh.go (hunk 3 above)
Comment (0xFelix) [reply tests/virtctl/ssh.go]: :+1:

File: tests/virtctl/ssh.go (hunk 4 above)
Comment (0xFelix) [reply tests/virtctl/ssh.go:51]: :+1:

File: tests/virtctl/scp.go (hunk 5 above)
Comment (0xFelix) [reply tests/virtctl/scp.go]: :+1:

File: tests/virtctl/scp.go (hunk 6 above)
Comment (0xFelix) [reply tests/virtctl/scp.go]: You are right, it is not required again.

File: tests/virtctl/ssh.go (hunk 7 above)
Comment (0xFelix) [reply tests/virtctl/ssh.go]: You are right, it is not required again.

**Comment by lyarwood** [comment]
/lgtm 

Thanks!

**Pull request merged**